File: app/schemas/transaction.py
from pydantic import BaseModel, field_validator
from decimal import Decimal
from datetime import date, datetime

class TransactionCreate(BaseModel):
    category_id: int
    amount: Decimal
    description: str | None = None
    transaction_date: date

class TransactionUpdate(BaseModel):
    category_id: int | None = None
    amount: Decimal | None = None
    description: str | None = None
    transaction_date: date | None = None

class TransactionResponse(BaseModel):
    id: int
    category_id: int
    amount: Decimal
    description: str | None = None
    transaction_date: date
    created_at: datetime

    class Config:
        from_attributes = True

class TransactionImportRowResult(BaseModel):
    row_number: int  # 1-indexed, counting the header row, so it matches what a spreadsheet shows
    status: str  # "imported" or "failed"
    error: str | None = None

class TransactionImportSummary(BaseModel):
    total_rows: int
    imported_count: int
    failed_count: int
    results: list[TransactionImportRowResult]

class CategorySuggestionRequest(BaseModel):
    description: str

class CategorySuggestionResponse(BaseModel):
    category_id: int | None = None
    category_name: str | None = None

File: frontend/src/types/transaction.ts
// Mirrors app/schemas/transaction.py.
// `amount` is kept as a string end-to-end on the frontend (never parsed into a
// JS `number` for editing/sending) so we never introduce floating-point
// rounding into a value the backend treats as an exact Decimal.

export interface TransactionResponse {
  id: number;
  category_id: number;
  amount: string;
  description: string | null;
  transaction_date: string; // YYYY-MM-DD
  created_at: string;
}

export interface TransactionCreatePayload {
  category_id: number;
  amount: string;
  description?: string | null;
  transaction_date: string;
}

export interface TransactionUpdatePayload {
  category_id?: number;
  amount?: string;
  description?: string | null;
  transaction_date?: string;
}

export interface TransactionFilters {
  category_id?: number;
  start_date?: string;
  end_date?: string;
  search?: string;
  skip?: number;
  limit?: number;
}

export interface TransactionImportRowResult {
  row_number: number;
  status: "imported" | "failed";
  error: string | null;
}

export interface TransactionImportSummary {
  total_rows: number;
  imported_count: number;
  failed_count: number;
  results: TransactionImportRowResult[];
}
